### ISSUE TITLE
Replace middleware with interceptor API

### ISSUE THREAD

**Issue by sagebind** [sagebind/isahc#42]
Currently middleware is behind the `middleware-api` feature flag, and is an implementation detail of cookie handling. We should refine the middleware API and then release it.

The current API is a bit limiting, and has no way of canceling a request or short-circuting a request. A useful and inriguing use of middleware would be to implement proper client-side caching of resources (like browsers) that respect ETags and conditional requests using a persistent backing store. As-is the middleware API doesn't let you implement something like this.

Here's a Playground link I shared in a Twiiter conversation for an idea on how the API could be improved: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4dfd3a8b0c1c306f175390170e8eff14

Here are some requirements and use-cases that the API should allow you to accomplish:

- Client-side caching
- Cookie handling using an async storage backend
- Returning mocked responses (#106)

**Comment by bluetech** [comment]
I'd like to suggest renaming Middleware -> Interceptor. For some reason, for me at least, "middleware" is associated with server-side, while "interceptor" is the same concept on the client-side.

A couple of examples:

- Angular: https://angular.io/api/common/http/HttpInterceptor
- OkHttp (Android): https://square.github.io/okhttp/interceptors/

**Comment by sagebind** [comment]
Proposed API:

```rust
pub trait Interceptor: Send {
    fn intercept_request(
        &mut self,
        request: Request<Body>,
    ) -> FutureObj<RequestAction> {
        RequestAction
    }

    fn intercept_response(
         &mut self,
         request: &Request<Body>,
         response: Response<Body>,
    ) -> FutureObj<Response<Body>> {
        future::ok(response).into()
    }
}

```

**Comment by sagebind** [comment]
Here's an initial proposal as to what the API could look like:

```rust
pub trait Interceptor: Sync {
    fn intercept_request(
        &mut self,
        request: Request<Body>,
    ) -> FutureObj<RequestAction> {
        future::ready(RequestAction::Continue(request)).into()
    }

    fn intercept_response(
         &mut self,
         request: &Request<Body>,
         response: Response<Body>,
    ) -> FutureObj<Response<Body>> {
        future::ready(response).into()
    }
}

pub enum RequestAction {
    /// The interceptor returned a modified request.
    Continue(Request<Body>),

    /// The interceptor interrupted the chain with a response.
    Halt(Response<Body>),
}
```

**Comment by sagebind** [comment]
Removing from 1.0 milestone, as this feature isn't necessary for a 1.0 release. It can always be added after without breaking changes if necessary.

**Comment by sagebind** [comment]
This is now available in the [0.9.11 release](https://github.com/sagebind/isahc/releases/tag/0.9.11) behind an `unstable-interceptors` feature. A separate issue was created to track stabilization: #247.
